### DISCUSSION TITLE
Building a Learned Sparse Index

### DISCUSSION THREAD

**Discussion by paulowoicho** [castorini/pyserini#1600] [Q&A]
Hi, I was looking at building a learned sparse index from some documents but I noticed there wasn't any documentation on it at https://github.com/castorini/pyserini/blob/master/docs/usage-index.md#building-a-sparse-vector-index. I found and read through https://github.com/castorini/pyserini/blob/master/docs/experiments-spladev2.md which helped me make some assumptions about the process that I wanted to check.

Say I have my collection in jsonlines format where every document has an `id` and `contents` field. Am I correct in assuming that I need to pretokenise the `contents` field and save that into a `vector` field before running:

```
python -m pyserini.index.lucene \
  --collection JsonVectorCollection \
  --input path/to/documents \
  --index path/to/index \
  --generator DefaultLuceneDocumentGenerator \
  --threads 12 \
  --impact --pretokenized
```

And per, https://github.com/castorini/pyserini/blob/master/docs/usage-index.md#building-a-dense-vector-index, can I tokenize my documents by running:

```
python -m pyserini.encode \
  input   --corpus path/to/documents \
          --fields text \
  output  --embeddings path/to/output/dir \
  encoder --encoder encoder_name \
          --fields text \
          --batch 32 \
          --fp16
```

Or could I simply do something like:
```
import transformers

tokenizer = transformers.AutoTokenizer.from_pretrained(encoder_name)

vocab_lookup = {v: k for k, v in tokenizer.vocab.items()}

def tokenize(text: str) -> Dict[str, int]:
    """Gets the vector for the given text."""
    input_ids = tokenizer(text)["input_ids"]
    return {vocab_lookup[input_id]: input_id for input_id in input_ids if vocab_lookup[input_id] != "[CLS]"}
```

Thanks!

**Comment by lintool** [comment]
Hi @paulowoicho - based on quick glance, this looks right to me.

@MXueguang any further comments?

Please let us know if you run into any specific issues.

Also, PRs to add to existing documentation would be greatly appreciated!

**Reply by paulowoicho** [reply]
Thanks, much appreciated!

The collection I'm indexing is quite large and is taking a few days as a result. I'll look into making a PR once I'm done indexing and testing it.
